system/lib/libc/emscripten_setjmp.c

```diff
@@ -0,0 +1,57 @@
+/*
+* Copyright 2020 The Emscripten Authors.  All rights reserved.
+* Emscripten is available under two separate licenses, the MIT license and the
+* University of Illinois/NCSA Open Source License.  Both these licenses can be
+* found in the LICENSE file.
+*/
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <setjmp.h>
+
+static uint32_t setjmpId = 0;
+
+struct TableEntry {
+  uint32_t id, label;
+};
+
+extern void setTempRet0(uint32_t value);
+
+struct TableEntry* saveSetjmp(uint32_t* env, uint32_t label, struct TableEntry* table, uint32_t size) {
+  // Not particularly fast: slow table lookup of setjmpId to label. But setjmp
+  // prevents relooping anyhow, so slowness is to be expected. And typical case
+  // is 1 setjmp per invocation, or less.
+  uint32_t i = 0;
+  setjmpId++;
+  *env = setjmpId;
+  while (i < size) {
+    if (table[i].id == 0) {
+      table[i].id = setjmpId;
+      table[i].label = label;
+      // prepare next slot
+      table[i + 1].id = 0;
+      setTempRet0(size);
+      return table;
+    }
+    i++;
+  }
+  // grow the table
+  size *= 2;
+  table = (struct TableEntry*)realloc(table, sizeof(struct TableEntry) * (size +1));
+  table = saveSetjmp(env, label, table, size);
+  setTempRet0(size); // FIXME: unneeded?
+  return table;
+}
+
+uint32_t testSetjmp(uint32_t id, struct TableEntry* table, uint32_t size) {
+  uint32_t i = 0, curr;
+  while (i < size) {
+    uint32_t curr = table[i].id;
+    if (curr == 0) break;
+    if (curr == id) {
+      return table[i].label;
+    }
+    i++;
+  }
+  return 0;
+}
```

Comment:
The symbols defined in `system/lib/compiler-rt/extras.c` fall into a similar catagory as these.  i.e. they are compiler support symbols which are used by the exception handling and sjlj code lowering passes.

I wonder if it makes sense for these to be in compiler-rt too?  Maybe in the same fiel?

Reply:
`extras.c` is upstream-only, so that would mean writing a second solution for fastcomp atm. maybe can leave that for after fastcomp is gone?

Reply:
I'm fine with this landing as is if you like.   But I'd also be OK with merging those two files that putting some parts under `ifdef __wasm__`?

Also would it makes sense to at least put both of them in compiler-rt?

Reply:
Ok, I pushed a commit now to move them to compiler-rt in a separate file, and get rid of the fastcomp-specific logic.